### PR TITLE
JP-3900: Tests, clean up, and code style for residual_fringe

### PR DESCRIPTION
**Pull request by melanieclarke**
Resolves [JP-3900](https://jira.stsci.edu/browse/JP-3900)

Add test coverage for residual_fringe and update code style to new standards.

While writing unit tests, I noticed some minor bugs, inefficiencies, and unused code, so I cleaned these out as well. None of these changes are likely to have any impact on the user -- this is just a maintainability improvement.

The truth files for the new regression test were generated with the code on the main branch, to make sure this PR does not introduce any regresssions.

## Tasks
- [ ] **request a review from someone specific**, to avoid making the maintainers review every PR
- [ ] add a build milestone, i.e. `Build 11.3` (use the [latest build](https://github.com/spacetelescope/jwst/milestones) if not sure)
- [ ] Does this PR change user-facing code / API? (if not, label with `no-changelog-entry-needed`)
  - [ ] write news fragment(s) in `changes/`: `echo "changed something" > changes/<PR#>.<changetype>.rst` (see below for change types) 
  - [ ] update or add relevant tests
  - [ ] update relevant docstrings and / or `docs/` page
  - [ ] [start a regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/jwst.yml) and include a link to the running job ([click here for instructions](https://github.com/spacetelescope/RegressionTests/blob/main/docs/running_regression_tests.md))
    - [ ] Do truth files need to be updated ("okified")?
      - [ ] **after the reviewer has approved these changes**, run `okify_regtests` to update the truth files
- [ ] if a JIRA ticket exists, [make sure it is resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

<details><summary>news fragment change types...</summary>

- ``changes/<PR#>.general.rst``: infrastructure or miscellaneous change
- ``changes/<PR#>.docs.rst``
- ``changes/<PR#>.stpipe.rst``
- ``changes/<PR#>.datamodels.rst``
- ``changes/<PR#>.scripts.rst``
- ``changes/<PR#>.set_telescope_pointing.rst``
- ``changes/<PR#>.pipeline.rst``

## stage 1
- ``changes/<PR#>.group_scale.rst``
- ``changes/<PR#>.dq_init.rst``
- ``changes/<PR#>.emicorr.rst``
- ``changes/<PR#>.saturation.rst``
- ``changes/<PR#>.ipc.rst``
- ``changes/<PR#>.firstframe.rst``
- ``changes/<PR#>.lastframe.rst``
- ``changes/<PR#>.reset.rst``
- ``changes/<PR#>.superbias.rst``
- ``changes/<PR#>.refpix.rst``
- ``changes/<PR#>.linearity.rst``
- ``changes/<PR#>.rscd.rst``
- ``changes/<PR#>.persistence.rst``
- ``changes/<PR#>.dark_current.rst``
- ``changes/<PR#>.charge_migration.rst``
- ``changes/<PR#>.jump.rst``
- ``changes/<PR#>.clean_flicker_noise.rst``
- ``changes/<PR#>.ramp_fitting.rst``
- ``changes/<PR#>.gain_scale.rst``

## stage 2
- ``changes/<PR#>.assign_wcs.rst``
- ``changes/<PR#>.badpix_selfcal.rst``
- ``changes/<PR#>.msaflagopen.rst``
- ``changes/<PR#>.nsclean.rst``
- ``changes/<PR#>.imprint.rst``
- ``changes/<PR#>.background.rst``
- ``changes/<PR#>.extract_2d.rst``
- ``changes/<PR#>.master_background.rst``
- ``changes/<PR#>.wavecorr.rst``
- ``changes/<PR#>.srctype.rst``
- ``changes/<PR#>.straylight.rst``
- ``changes/<PR#>.wfss_contam.rst``
- ``changes/<PR#>.flatfield.rst``
- ``changes/<PR#>.fringe.rst``
- ``changes/<PR#>.pathloss.rst``
- ``changes/<PR#>.barshadow.rst``
- ``changes/<PR#>.photom.rst``
- ``changes/<PR#>.pixel_replace.rst``
- ``changes/<PR#>.resample_spec.rst``
- ``changes/<PR#>.residual_fringe.rst``
- ``changes/<PR#>.cube_build.rst``
- ``changes/<PR#>.extract_1d.rst``
- ``changes/<PR#>.resample.rst``

## stage 3
- ``changes/<PR#>.assign_mtwcs.rst``
- ``changes/<PR#>.mrs_imatch.rst``
- ``changes/<PR#>.tweakreg.rst``
- ``changes/<PR#>.skymatch.rst``
- ``changes/<PR#>.exp_to_source.rst``
- ``changes/<PR#>.outlier_detection.rst``
- ``changes/<PR#>.tso_photometry.rst``
- ``changes/<PR#>.stack_refs.rst``
- ``changes/<PR#>.align_refs.rst``
- ``changes/<PR#>.klip.rst``
- ``changes/<PR#>.spectral_leak.rst``
- ``changes/<PR#>.source_catalog.rst``
- ``changes/<PR#>.combine_1d.rst``
- ``changes/<PR#>.ami.rst``

## other
- ``changes/<PR#>.wfs_combine.rst``
- ``changes/<PR#>.white_light.rst``
- ``changes/<PR#>.cube_skymatch.rst``
- ``changes/<PR#>.engdb_tools.rst``
- ``changes/<PR#>.guider_cds.rst``
</details>

### CODE REVIEW

**Comment by melanieclarke** [comment]
Regression tests here:
https://github.com/spacetelescope/RegressionTests/actions/runs/13662806122

**Comment by melanieclarke** [comment]
FYI @drlaw1558 - while implementing tests for residual_fringe, I noticed some TODOs in the code, to remove the contrast and quality calculations and output table.  I think they were leftover from some of your work on implementing the 1D residual fringe correction.  I went ahead and removed the unused code here.

**Review by emolter — COMMENTED** [review]
Looks much cleaner Melanie, thanks for this!  I added a few small comments.

File: jwst/residual_fringe/tests/test_residual_fringe.py [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+        xrange_channel = np.array([[0, 10]])
+        slice_x_ranges = np.array([[301, 0, 10]])
+        all_slice_masks = np.ones((1, 1024, 10))
+        return slices_in_channel, xrange_channel, slice_x_ranges, all_slice_masks
+
+    monkeypatch.setattr(utils, "slice_info", one_slice)
+
+
+@pytest.fixture()
+def module_log_watcher(monkeypatch):
```
[the remark this answers is not in the document]
Comment (melanieclarke) [reply jwst/residual_fringe/tests/test_residual_fringe.py:124]: That's a great idea.  I started adding these for specific purposes in different modules, but it's now become a pattern I use every time I'm building out unit tests.  I'll file a ticket to get that in in a separate PR, so I can modify all the places it's used at once.

File: jwst/residual_fringe/tests/test_utils.py
[the remark this answers is not in the document]
Comment (melanieclarke) [reply jwst/residual_fringe/tests/test_utils.py:1]: The utils module is 87% covered, with this test + the others that test outputs at a higher level.  I added an explicit test for the `slice_info` function because it's mocked in all the other tests for speed.  The remainder of the uncovered lines are mostly edge case handling that I think are not worth the effort to get covered right now.

**Comment by melanieclarke** [comment]
Thanks for reviewing @emolter!  I'll get those fixes in shortly.

**Review by kmacdonald-stsci — APPROVED** [review]

File: jwst/residual_fringe/tests/test_residual_fringe.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (emolter) [reply jwst/residual_fringe/tests/test_residual_fringe.py:124]: that sounds like a great plan

File: jwst/residual_fringe/residual_fringe.py [hunk 2]
```diff
@@ -144,326 +169,348 @@ def do_correction(self):
...
-            del slice_x_ranges, all_slice_masks, slices_in_band, wave_map  # end of channel
-            log.info('Number of columns corrected for channel {}'.format(num_corrected))
+                    # use the error array to get col snr, used to remove noisy pixels
+                    col_snr = self.model.data[:, col] / self.model.err[:, col]
+
+                    # Do some checks on column to make sure there is
+                    # reasonable signal. If the SNR < min_snr (CDP), pass
+                    n = len(test_flux)
+                    signal = np.nanmean(test_flux)
+                    noise = 0.6052697 * np.nanmedian(
```
[the remark this answers is not in the document]
Comment (melanieclarke) [reply jwst/residual_fringe/residual_fringe.py:270]: This was a fun little side trip.  It appears to derive from a spectroscopic signal-to-noise calculation also used in specutils, documented here: https://archive.stsci.edu/vodocs/der_snr.pdf

I'll put it in a constant at the top of the file and add the reference in a comment.

**Comment by drlaw1558** [comment]
> FYI @drlaw1558 - while implementing tests for residual_fringe, I noticed some TODOs in the code, to remove the contrast and quality calculations and output table. I think they were leftover from some of your work on residual fringe correction. I went ahead and removed the unused code here.

Thanks- given its heritage some of this code is pretty convoluted.

File: jwst/residual_fringe/tests/test_residual_fringe.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (melanieclarke) [reply jwst/residual_fringe/tests/test_residual_fringe.py:124]: I filed a GitHub issue to track, since it's an entirely internal technical issue: #9244

**Comment by melanieclarke** [comment]
> Thanks- given its heritage some of this code is pretty convoluted.

Yes, it looks like it could use a more general assessment and refactor.  In particular, it would be nice to see if we can either replace or reimplement the BayesicFitting dependency with something less fragile.

File: jwst/residual_fringe/tests/test_residual_fringe.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (melanieclarke) [reply jwst/residual_fringe/tests/test_residual_fringe.py]: Local implementations are now replaced with the new general fixture.

**Review by jemorrison — APPROVED** [review]
Looks good to me. Many more units tests are good addition.
